// solid-auth-client/lib/api.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.logout = exports.currentSession = exports.popupLogin = exports.login = exports.fetch = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var firstSession = function () {
  var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(storage, authFns) {
    var _session;

    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            if (!(authFns.length === 0)) {
              _context.next = 2;
              break;
            }

            return _context.abrupt('return', null);

          case 2:
            _context.prev = 2;
            _context.next = 5;
            return authFns[0]();

          case 5:
            _session = _context.sent;

            if (!_session) {
              _context.next = 8;
              break;
            }

            return _context.abrupt('return', (0, _session2.saveSession)(storage)(_session));

          case 8:
            _context.next = 13;
            break;

          case 10:
            _context.prev = 10;
            _context.t0 = _context['catch'](2);

            console.error(_context.t0);

          case 13:
            return _context.abrupt('return', firstSession(storage, authFns.slice(1)));

          case 14:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[2, 10]]);
  }));

  return function firstSession(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var login = exports.login = function () {
  var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(idp, options) {
    var webIdTlsSession, webIdOidcLoginRedirectFn;
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            options = (0, _extends3.default)({}, defaultLoginOptions(), options);
            _context2.next = 3;
            return WebIdTls.login(idp);

          case 3:
            webIdTlsSession = _context2.sent;

            if (!webIdTlsSession) {
              _context2.next = 6;
              break;
            }

            return _context2.abrupt('return', (0, _session2.saveSession)(options.storage)(webIdTlsSession));

          case 6:
            _context2.next = 8;
            return WebIdOidc.login(idp, options);

          case 8:
            webIdOidcLoginRedirectFn = _context2.sent;
            return _context2.abrupt('return', webIdOidcLoginRedirectFn);

          case 10:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function login(_x3, _x4) {
    return _ref2.apply(this, arguments);
  };
}();

var popupLogin = exports.popupLogin = function () {
  var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(options) {
    var childWindow, session;
    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            if (options.popupUri) {
              _context3.next = 2;
              break;
            }

            throw new Error('Must provide options.popupUri');

          case 2:
            if (!options.callbackUri) {
              options.callbackUri = options.popupUri;
            }
            options = (0, _extends3.default)({}, defaultLoginOptions(), options);
            childWindow = (0, _popup.openIdpSelector)(options);
            _context3.next = 7;
            return (0, _popup.startPopupServer)(options.storage, childWindow, options);

          case 7:
            session = _context3.sent;
            return _context3.abrupt('return', session);

          case 9:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function popupLogin(_x5) {
    return _ref3.apply(this, arguments);
  };
}();

var currentSession = exports.currentSession = function () {
  var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4() {
    var storage = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _storage.defaultStorage)();
    var session;
    return _regenerator2.default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return (0, _session2.getSession)(storage);

          case 2:
            session = _context4.sent;

            if (!session) {
              _context4.next = 5;
              break;
            }

            return _context4.abrupt('return', session);

          case 5:
            return _context4.abrupt('return', firstSession(storage, [WebIdOidc.currentSession.bind(null, storage)]));

          case 6:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this);
  }));

  return function currentSession() {
    return _ref4.apply(this, arguments);
  };
}();

var logout = exports.logout = function () {
  var _ref5 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5() {
    var storage = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _storage.defaultStorage)();
    var session;
    return _regenerator2.default.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return (0, _session2.getSession)(storage);

          case 2:
            session = _context5.sent;

            if (session) {
              _context5.next = 5;
              break;
            }

            return _context5.abrupt('return');

          case 5:
            _context5.t0 = session.authType;
            _context5.next = _context5.t0 === 'WebID-OIDC' ? 8 : _context5.t0 === 'WebID-TLS' ? 18 : 18;
            break;

          case 8:
            _context5.prev = 8;
            _context5.next = 11;
            return WebIdOidc.logout(storage);

          case 11:
            _context5.next = 17;
            break;

          case 13:
            _context5.prev = 13;
            _context5.t1 = _context5['catch'](8);

            console.warn('Error logging out:');
            console.error(_context5.t1);

          case 17:
            return _context5.abrupt('break', 19);

          case 18:
            return _context5.abrupt('break', 19);

          case 19:
            return _context5.abrupt('return', (0, _session2.clearSession)(storage));

          case 20:
          case 'end':
            return _context5.stop();
        }
      }
    }, _callee5, this, [[8, 13]]);
  }));

  return function logout() {
    return _ref5.apply(this, arguments);
  };
}();

var _authnFetch = require('./authn-fetch');

var _popup = require('./popup');

var _session2 = require('./session');

var _storage = require('./storage');

var _urlUtil = require('./url-util');

var _webidTls = require('./webid-tls');

var WebIdTls = _interopRequireWildcard(_webidTls);

var _webidOidc = require('./webid-oidc');

var WebIdOidc = _interopRequireWildcard(_webidOidc);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultLoginOptions = function defaultLoginOptions() {
  var url = (0, _urlUtil.currentUrlNoParams)();
  return {
    callbackUri: url ? url.split('#')[0] : null,
    popupUri: null,
    storage: (0, _storage.defaultStorage)()
  };
};
/* global RequestInfo, Response */
var fetch = exports.fetch = function fetch(url, options) {
  return (0, _authnFetch.authnFetch)((0, _storage.defaultStorage)())(url, options);
};
